README.md
plugin.video.cc2tv
==================

XBMC-AddOn für den Zugriff auf die Video-/Audiopodcasts von CC2tv (ehemals Computer Club 2).


changelog.txt
10.02.2017 - v1.1.1
	- Fork des Plugins für CC2tv

22.06.2013 - v1.1.0
	- Audiopodcast hinzugefügt

25.05.2013 - v1.0.0
	Veröffentlichung des AddOns


tests/test_Parser.py
# *-* coding=utf8 *-*
import os
from resources.lib.Parser import Parser

class Test_Parser(object):
    
    def test__init__(self):
        parser = _get_parser_for_test_feed()
        # prüfe ob korrekt initialisiert wurde
        assert parser._url == _get_feed()
        assert parser._reobj_folge.pattern == _get_folge_pattern()
        assert parser._reobj_thema.pattern == _get_thema_pattern()
        
    def test_get_data(self):
        parser = _get_parser_for_test_feed()
        #prüfe ob die daten korrekt ermittelt wurden
        feed_data = parser.get_data()
        assert feed_data[0][0] == u"108. Folge"
        assert feed_data[0][1] == u"Raspberry PI \x96 ein erster Eindruck. "
        assert feed_data[0][2] == u"http://cczwei.mirror.speedpartner.de/cc2tv/CC2_108.mp4"
        
    def test__get_folge(self):
        parser = _get_parser_for_test_feed()
        assert parser._get_folge(u"CC2-NRWTV - 108. Folge") == "108. Folge"
        
    def test__get_thema(self):
        parser = _get_parser_for_test_feed()
        desc = u"Heute: CC2: Raspberry PI  ein erster Eindruck. "
        assert parser._get_thema(desc) == u"Raspberry PI \x96 ein erster Eindruck. "
        
def _get_parser_for_test_feed():
    return Parser(_get_feed(), _get_folge_pattern(), _get_thema_pattern() )

def _get_feed():
    return os.path.join(os.path.dirname(__file__), "data/feed.xhtml")

def _get_folge_pattern():
    return "- (\d+. Folge)"

def _get_thema_pattern():
    return "Heute: CC2: (.*)"
    


tests/test_RssHandler.py
# *-* coding=utf8 *-*
from resources.lib.RssHandler import RssHandler

class Test_RssHandler(object):

    def test__init__(self):
        rsshandler = RssHandler()
        assert ["","","",""] == _get_datafields(rsshandler)
    
    def test_startElement(self):
        # prüfe ob RssHandler korrekt mit Daten gefüllt werden kann
        rsshandler = _get_rsshandler_with_data()
        assert _get_datafields(rsshandler) == ["a","b","c","d"]

        # prüfe ob diese Schlüsselworte kein Reset ausführen
        for key in ("title","pubDate","description","enclosure"):
            rsshandler = _get_rsshandler_with_data()
            rsshandler.startElement(key, {"url":"d"})
            assert _get_datafields(rsshandler) == ["a","b","c","d"]

        # prüfe ob "item" ein Reset durchführt
        rsshandler = _get_rsshandler_with_data()
        rsshandler.startElement("item", {"url":"d"})
        assert _get_datafields(rsshandler) == ["","","",""]

        # prüfe ob die folgende Schlüsselworte das Feld "_aktiv" ändern
        for key in ("title","pubDate","description"):
            rsshandler = _get_rsshandler_with_data()
            rsshandler.startElement(key, {"url":"d"})
            assert rsshandler._aktiv == key

        # prüfe ob die folgenden Schlüsselworte das Feld "_aktiv" nicht ändern
        for key in ("item","enclosure"):
            rsshandler = _get_rsshandler_with_data()
            rsshandler.startElement(key, {"url":"d"})
            assert rsshandler._aktiv != key
            
        # prüfe ob "enclosure" korrekt gesetzt wird
        rsshandler = _get_rsshandler_with_data()
        rsshandler.startElement("enclosure", {"url":"/path/to/nowhere"})
        assert rsshandler._enclosure == "/path/to/nowhere"
        
    def test_endElement(self):
        # prüfe ob Feld "result" vorm Start leer ist
        rsshandler = _get_rsshandler_with_data()
        assert rsshandler.result == []
        
        # prüfe ob bei "item" alle Datenfelder in "result" übernommen werden
        rsshandler.endElement("item")
        assert rsshandler.result == [("a","b","c","d")]
        
        # prüfe ob die folgenden Schlüsselwörter das Feld "_aktiv" zurücksetzen
        for key in ("title","pubDate","description","enclosure"):
            rsshandler = _get_rsshandler_with_data()
            rsshandler._aktiv = key
            rsshandler.endElement(key)
            assert rsshandler._aktiv == None
            
        # prüfe ob "item" das Feld "_aktiv" nicht zurücksetzt
        rsshandler = _get_rsshandler_with_data()
        feldinhalt = "bleibt unangetastet!"
        rsshandler._aktiv = feldinhalt
        rsshandler.endElement("item")
        assert rsshandler._aktiv == feldinhalt

    def test_characters(self):
        # prüfe ob Daten richtig zugeordnet werden
        rsshandler = RssHandler()
        rsshandler._aktiv = "title"
        rsshandler.characters("testdaten title")
        assert rsshandler._title == "testdaten title"
        
        rsshandler._aktiv = "pubDate"
        rsshandler.characters("testdaten pubDate")
        assert rsshandler._pubDate == "testdaten pubDate"
        
        rsshandler._aktiv = "description"
        rsshandler.characters("testdaten description")
        assert rsshandler._description == "testdaten description"
        

def _get_rsshandler_with_data():
    rsshandler = RssHandler()
    rsshandler._title = "a"
    rsshandler._pubDate = "b"
    rsshandler._description = "c"
    rsshandler._enclosure = "d"
    return rsshandler
    
def _get_datafields(rsshandler):
    return [
        rsshandler._title,
        rsshandler._pubDate,
        rsshandler._description,
        rsshandler._enclosure
    ]    
   


addon.py
# -*- coding=utf8 -*-
#******************************************************************************
# addon.py
#------------------------------------------------------------------------------
#
# Copyright (c) 2013 LivingOn <LivingOn@xmail.net>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
#******************************************************************************
import sys
import urllib
import xbmcgui
import xbmcaddon
import xbmcplugin

from resources.lib.Parser import Parser

class CCZwei(object):
    
    PLUGIN_NAME = "plugin.video.cc2tv"
    
    _plugin_id      = None
    _addon          = None
    _rss_video_feed = None
    _rss_audio_feed = None
    _regex_folge    = None
    _regex_thema    = None
    
    def __init__(self):
        self._register_addon()
        self._load_settings()
        self._add_directory_items()
        
    def _register_addon(self):
        self._plugin_id = int(sys.argv[1])
        self._addon = xbmcaddon.Addon(id = self.PLUGIN_NAME)

    def _load_settings(self):
        self._rss_video_feed = xbmcplugin.getSetting(self._plugin_id, "rss_video_feed")
        self._rss_audio_feed = xbmcplugin.getSetting(self._plugin_id, "rss_audio_feed")
        self._regex_folge = r"%s" % xbmcplugin.getSetting(self._plugin_id, "regex_folge")
        self._regex_thema = r"%s" % xbmcplugin.getSetting(self._plugin_id, "regex_thema")

    def _add_directory_items(self):
        if not sys.argv[2]:
            self._create_submenues()
        else:
            self._read_feed_data_into_directory(sys.argv[2])

    def _create_submenues(self):
        self._add_item_to_directory("Videopodcast", None, True)
        self._add_item_to_directory("Audiopodcast", None, True)
        xbmcplugin.endOfDirectory(self._plugin_id)

    def _read_feed_data_into_directory(self, mode):
        if mode == "?mode=Videopodcast":
            rss_feed = self._rss_video_feed
        else:
            rss_feed = self._rss_audio_feed
        feed_data = Parser(rss_feed, self._regex_folge, self._regex_thema).get_data()
        if feed_data:
            for (folge, thema, stream_url) in feed_data:
                self._add_item_to_directory("%s - %s" % (folge, thema), stream_url)
            xbmcplugin.endOfDirectory(self._plugin_id)
        else:
            title = self._addon.getLocalizedString(30110)
            msg = self._addon.getLocalizedString(30120)
            xbmcgui.Dialog().ok(title, msg)

    def _add_item_to_directory(self, title, url, isfolder=False):
        if isfolder:
            url = sys.argv[0] + "?" + urllib.urlencode({'mode' : title})
        item = xbmcgui.ListItem(title)
        xbmcplugin.addDirectoryItem(self._plugin_id, url, item, isfolder)
            
if __name__ == "__main__":
    CCZwei()


build_ziparchiv.py
#!/usr/bin/env python
from zipfile import ZipFile, ZIP_DEFLATED

class Build(object):
    
    NAME = "plugin.video.cc2tv"
    FILES = [
        "addon.py",
        "addon.xml",
        "changelog.txt",
#        "icon.png",
        "LICENSE.TXT",
        "resources/__init__.py",
        "resources/settings.xml",
        "resources/language/English/strings.xml",
        "resources/language/German/strings.xml",
        "resources/lib/__init__.py",
        "resources/lib/Parser.py",
        "resources/lib/RssHandler.py"
    ]
    
    def create_ziparchiv(self, path):
        print("create ziparchiv %s" % path)
        ziparchiv = ZipFile(path, "w")
        count = 0
        for f in self.FILES:
            print("   add %s" % f)
            count += 1
            ziparchiv.write(f, "%s/%s" % (self.NAME,f), ZIP_DEFLATED)
        ziparchiv.close()
        print("%d files added" % count)
            
    
if __name__ == "__main__":
    build = Build()
    build.create_ziparchiv("%s.zip" % build.NAME)    
        


resources/lib/Parser.py
# -*- coding=utf8 -*-
#******************************************************************************
# Parser.py
#------------------------------------------------------------------------------
#
# Copyright (c) 2013 LivingOn <LivingOn@xmail.net>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
#******************************************************************************
import re
import xml.sax as sax

from resources.lib.RssHandler import RssHandler

class Parser(object):
    
    _url         = None
    _reobj_folge = None
    _reobj_thema = None
    
    def __init__(self, url, regex_folge, regex_thema):
        self._url = url
        self._reobj_folge = re.compile(regex_folge)
        self._reobj_thema = re.compile(regex_thema)
        
    def get_data(self):
        handler = RssHandler()
        try: 
            parser = sax.make_parser() 
            parser.setContentHandler(handler)
            parser.parse(self._url) 
            result = []
            for (title, dummy_date, desc, url) in handler.result:
                result.append((
                    self._get_folge(title), 
                    self._get_thema(desc), 
                    url
                ))
        except Exception:
            result = None
        return result
    
    def _get_folge(self, title):
        return self._filter_text(self._reobj_folge, title)
    
    def _get_thema(self, desc):
        return self._filter_text(self._reobj_thema, desc)
    
    def _filter_text(self, reobj, text):
        try:
            result = reobj.search(text).group(1)
        except: 
            result = text
        return result 


resources/lib/RssHandler.py
# -*- coding=utf8 -*-
#******************************************************************************
# RssHandler.py
#------------------------------------------------------------------------------
#
# Copyright (c) 2009, 2013 LivingOn <LivingOn@xmail.net>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
#******************************************************************************
import xml.sax as sax

class RssHandler(sax.handler.ContentHandler):

    result       = None
    _aktiv       = None
    _title       = None
    _pubDate     = None
    _description = None
    _enclosure   = None

    def __init__(self): 
        self.result = []
        self._aktiv = None 
        self._resetVariables()
    
    def startElement(self, name, attrs): 
        if name == "item": 
            self._resetVariables()
        elif name == "title" or name == "pubDate" or name == "description": 
            self._aktiv = name
        elif name == "enclosure":
            self._enclosure = attrs["url"]

    def endElement(self, name):
        if name == "item": 
            self.result.append( 
               (self._title, self._pubDate, self._description, self._enclosure))
        elif name == "title" or name == "pubDate" or \
                                name == "description" or name == "enclosure": 
            self._aktiv = None

    def characters(self, content): 
        if self._aktiv == "title": 
            self._title += content 
        elif self._aktiv == "pubDate": 
            self._pubDate += content
        elif self._aktiv == "description": 
            if self._description == "":
                self._description += content

    def _resetVariables(self):
        self._title = "" 
        self._pubDate = ""
        self._description = ""
        self._enclosure = ""
